core/crypto/crypto.cpp

```diff
@@ -82,6 +99,27 @@ void Crypto::load_default_certificates(String p_path) {
 	}
 }
 
+// Compares two HMACS for equality without leaking timing information in order to prevent timing attakcs.
+// @see: https://paragonie.com/blog/2015/11/preventing-timing-attacks-on-string-comparison-with-double-hmac-strategy
+bool Crypto::hmac_equals(PackedByteArray p_hm1, PackedByteArray p_hm2) {
+	const uint8_t *hm1 = p_hm1.ptr();
+	const uint8_t *hm2 = p_hm2.ptr();
+	int hm1_len = p_hm1.size();
+	int hm2_len = p_hm2.size();
+	// If the lengths are different then nothing else matters.
+	if (hm1_len != hm2_len) {
+		return false;
+	}
+
+	uint8_t v = 0;
+	for (int i = 0; i < hm1_len; i++) {
+		v |= hm1[i] ^ hm2[i];
+	}
+	// This is an extra precaution as to not leak timing information
+	// when comparing the byte to 0 either.
+	return bool(int((uint32_t(v ^ 0) - 1) >> 31));
```

Comment:
The reference implementation I used worked in 32-bit ints, and I wasn't sure if it was required for this comparison, but thinking about it more, it's probably fine to reduce it to a byte and operate there. Will adjust and try it.

Reply:
Also, the reason for the bool cast is because I like to be explicit with my types. No other reason.

Reply:
@Faless looks like it actually doesn't work with your alternative. It did point out one small bug though. I actually meant to use `int64_t` instead of `int` so that it works correctly on both 64-bit and 32-bit compilers.

Reply:
So thinking about it more, something bigger than 8-bits is needed (could also be 16-bit, doesn't really matter), since the algorithm requires you to shift beyond the 8-bits of a byte. `v` is always going to end up being 0 or 255. In the case that `v = 0` at the end of the loop (i.e. the byte arrays are the same), then `v ^ 0 = 0`. If the result is a byte, then when you subtract 1 from it, you get `b11111111`. If it's a uint32, it's `b11111111111111111111111111111111`. So when we shift past the bounds of a byte, it's either going to be 1 or 0. It's a good obfuscation for sure, but looking deeper at different implementations, it seems that the Go implementation is currently the only major language that implements it this way. All the other major languages that I looked at seem to just `return v== 0`. For now, I'll simplify it as such until I have more time to more deeply look into this.